tuf/tuf.go

```diff
@@ -294,114 +257,122 @@ func (tr *Repo) GetAllLoadedRoles() []*data.Role {
 	}
-	parent := path.Dir(role.Name)
+	parent := path.Dir(roleName)
 
 	if err := tr.VerifyCanSign(parent); err != nil {
 		return err
 	}
 
 	// check the parent role's metadata
-	p, ok := tr.Targets[parent]
+	_, ok := tr.Targets[parent]
 	if !ok { // the parent targetfile may not exist yet - if not, then create it
 		var err error
-		p, err = tr.InitTargets(parent)
+		_, err = tr.InitTargets(parent)
 		if err != nil {
 			return err
 		}
 	}
-
-	for _, k := range keys {
-		if !utils.StrSliceContains(role.KeyIDs, k.ID()) {
-			role.KeyIDs = append(role.KeyIDs, k.ID())
+	// Walk to parent, and either set or update this delegation
+	// Ensure all updates are valid
+	delegationUpdateVisitor := func(tgt *data.SignedTargets, validRole data.DelegationRole) interface{} {
+		var err error
+		// Validate the changes underneath this restricted validRole for adding paths, reject invalid path additions
+		if len(changes.AddPaths) != len(data.RestrictDelegationPathPrefixes(validRole.Paths, changes.AddPaths)) {
+			return data.ErrInvalidRole{Role: roleName, Reason: "invalid paths to add to role"}
+		}
+		// Try to find the delegation and amend it using our changelist
+		var delgRole *data.Role
+		for _, role := range tgt.Signed.Delegations.Roles {
+			if role.Name == roleName {
+				// Make a copy and operate on this role until we validate the changes
+				keyIDCopy := make([]string, len(role.KeyIDs))
+				copy(keyIDCopy, role.KeyIDs)
+				pathsCopy := make([]string, len(role.Paths))
+				copy(pathsCopy, role.Paths)
+				delgRole = &data.Role{
+					RootRole: data.RootRole{
+						KeyIDs:    keyIDCopy,
+						Threshold: role.Threshold,
+					},
+					Name:  role.Name,
+					Paths: pathsCopy,
+				}
+				delgRole.RemovePaths(changes.RemovePaths)
+				delgRole.AddPaths(changes.AddPaths)
+				if changes.ClearAllPaths {
+					delgRole.Paths = []string{}
+				}
+				delgRole.RemoveKeys(changes.RemoveKeys)
+				break
+			}
 		}
-		p.Signed.Delegations.Keys[k.ID()] = k
-	}
-
-	// if the role has fewer keys than the threshold, it
-	// will never be able to create a valid targets file
-	// and should be considered invalid.
-	if len(role.KeyIDs) < role.Threshold {
-		return data.ErrInvalidRole{Role: role.Name, Reason: "insufficient keys to meet threshold"}
+		// We didn't find the role earlier, so create it
+		if delgRole == nil {
+			delgRole, err = changes.ToNewRole(roleName)
+			if err != nil {
+				return err
+			}
+		}
+		// Add the key IDs to the role and the keys themselves to the parent
+		for _, k := range changes.AddKeys {
+			if !utils.StrSliceContains(delgRole.KeyIDs, k.ID()) {
+				delgRole.KeyIDs = append(delgRole.KeyIDs, k.ID())
+			}
+		}
+		// Make sure we have a valid role still
+		if len(delgRole.KeyIDs) < delgRole.Threshold {
+			return data.ErrInvalidRole{Role: roleName, Reason: "insufficient keys to meet threshold"}
+		}
+		// If valid, we should edit the actual SignedTargets to match our copy
```

Comment:
Can we add a comment in CAPS just above this noting that the remainder of this closure is not allowed to error as it is performing final modifications to the SignedTargets object and anything that might error _must_ be done first.